src/compiler/binder.ts

```diff
@@ -791,6 +800,13 @@ namespace ts {
                     bindPostfixUnaryExpressionFlow(<PostfixUnaryExpression>node);
                     break;
                 case SyntaxKind.BinaryExpression:
+                    if (isDestructuringAssignment(node)) {
+                        // Carry over whether we are in an assignment pattern to
+                        // binary expressions that could actually be an initializer
+                        inAssignmentPattern = saveInAssignmentPattern;
+                        bindDestructuringAssignmentFlow(node);
+                        return;

```

Comment:
You probably don't want to *return* here, and should instead just pop off the work item, as if you have something like `{ x = 2 } = y = {x: 0}` it's going to fail the bind the last half of the outer expression here, no?